### PR TITLE
rootless: fix return value handling

### PR DESCRIPTION
**Pull request by eriksjolund**
Fixes: https://github.com/containers/podman/issues/15927

Signed-off-by: Erik Sjölund <erik.sjolund@gmail.com>

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: pkg/rootless/rootless_linux.c [hunk 1]
```diff
@@ -506,15 +506,15 @@ create_pause_process (const char *pause_pid_file_path, char **argv)
...
+      int r, r2;
 
       close (p[1]);
       /* Block until we write the pid file.  */
       r = TEMP_FAILURE_RETRY (read (p[0], &b, 1));
       close (p[0]);
 
-      r = reexec_in_user_namespace_wait (pid, 0);
-      if (r != 0)
+      r2 = reexec_in_user_namespace_wait (pid, 0);
```
Comment (rhatdan) [inline pkg/rootless/rootless_linux.c:516]: This looks very different in the previous code reexec_in_user_namespace_wait was expected to succeed if it returned a 1, no you are failing it if returns a 1?

@giuseppe PTAL

**Comment by eriksjolund** [comment]
Maybe it should be something like this instead?

``` diff
diff --git a/pkg/rootless/rootless_linux.c b/pkg/rootless/rootless_linux.c
index fb22ed221..fb52a98c1 100644
--- a/pkg/rootless/rootless_linux.c
+++ b/pkg/rootless/rootless_linux.c
@@ -512,12 +512,12 @@ create_pause_process (const char *pause_pid_file_path, char **argv)
       /* Block until we write the pid file.  */
       r = TEMP_FAILURE_RETRY (read (p[0], &b, 1));
       close (p[0]);
+      if (r != 1 || b != '0') {
+        return -1;
+      }
 
       r = reexec_in_user_namespace_wait (pid, 0);
-      if (r != 0)
-       return -1;
-
-      return r == 1 && b == '0' ? 0 : -1;
+      return (r == 0) ? 0 : -1; 
     }
   else
     {
```

**Comment by eriksjolund** [comment]
In any case the current main branch version seems to have a bug here
https://github.com/containers/podman/blob/8656ffa5638dc25ee2ddeb817a5f7d5ae3493154/pkg/rootless/rootless_linux.c#L517-L520
The four lines will always result in
```
return -1;
```

That code came from
https://github.com/containers/podman/pull/14638/commits/99258703a7c593c2717438928cd9c9052e50c26b

**Comment by rhatdan** [comment]
Thanks, then your code looks correct, but I am not sure why @giuseppe did not totally break everything.
LGTM

**Comment by eriksjolund** [comment]
I guess @giuseppe will know how to make a proper fix. (I don't know this code so well)

File: pkg/rootless/rootless_linux.c (hunk 1 above)
Comment (giuseppe) [reply pkg/rootless/rootless_linux.c:516]: this looks indeed like a regression that was introduced with 99258703a7c593c2717438928cd9c9052e50c26b as it overwrites the value we care about.

**Comment by rhatdan** [comment]
@mheon This needs to get into 4.3 release.

**Pull request merged**